### ISSUE TITLE
What happens during training with HuggingFace models in eval mode?

### ISSUE THREAD

**Issue by StevenSong** [Lightning-AI/pytorch-lightning#20105]
### Bug description

I'm using huggingface models with lightning, a use case that seems to be part of the [lightning examples](https://lightning.ai/lightning-ai/studios/finetune-hugging-face-bert-with-pytorch-lightning). However, a subtlety of the huggingface API is that models are loaded by default in `eval` mode, as [documented](https://huggingface.co/docs/transformers/en/main_classes/model#transformers.PreTrainedModel.from_pretrained). There is currently no check that within `LightningModule.training_step`, models are in `train` mode. If a `LightningModule.validation_step` is defined, lightning will put the model in `train` mode at the end of the validation step, however this is only after the first training epoch completes. If no validation step is defined, all of training is done in `eval` mode. The below example only crashes because of the assertion at the start of `training_step`.

While training in eval mode is not a bug that leads to a crash and is technically allowed when using pure pytorch, in lightning, where the details of training a pytorch model are supposed to be abstracted, enforcing such a constraint feels like it ought to be the default.

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

```python
import os

import torch
from lightning.pytorch import LightningModule, Trainer
from torch.utils.data import DataLoader, Dataset
from transformers import AutoModelForCausalLM, AutoTokenizer

hf_model = AutoModelForCausalLM.from_pretrained("openai-community/gpt2")
tok = AutoTokenizer.from_pretrained("openai-community/gpt2")

class MyDataset(Dataset):
    def __getitem__(self, index):
        return tok("hello world", return_tensors="pt")

    def __len__(self):
        return 1


class MyModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.model = hf_model

    def training_step(self, batch, batch_idx):
        assert self.model.training
        loss = self.model(**batch).loss
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.model.parameters(), lr=0.1)


def run():
    train_data = DataLoader(MyDataset(), batch_size=1)

    model = MyModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=1,
        num_sanity_val_steps=0,
        max_epochs=1,
        enable_model_summary=False,
    )
    trainer.fit(model, train_dataloaders=train_data)


if __name__ == "__main__":
    run()
```

### Error messages and logs

Rather insidious as everything runs without crashing, I only discovered this because I was making a hybrid model with an RNN where cudnn RNN requires the model be in train mode, all my other models without RNN components trained to completion

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version: 2.2.4
#- PyTorch Version: 2.1.1+cu121
#- Python version: 3.10.13
#- OS: Ubuntu 20.04.6 LTS
#- CUDA/cuDNN version: 12.1
#- GPU models and configuration: NVIDIA A100 80GB PCIe
#- How you installed Lightning(`conda`, `pip`, source): pip
```

</details>

**Comment by awaelchli** [comment]
Hey @StevenSong

I changed this in #18951 (Lightning 2.2) so that now Lightning will remember in which training mode the model was before entering in validation mode, so that when validation is over, Lightning can restore exactly the previous mode it was in. This should help everyone who finetunes models 🎉 

However, as you noticed, this also means for HuggingFace pretrained models, they are defaulting to be used for inference. So you need to call `.train()` on them.

Can you make this modification and check please?

```py
hf_model = AutoModelForCausalLM.from_pretrained("openai-community/gpt2").train()
```

Another user proposed to show a warning: #19820 (which I am against but feel free to join the discussion 🚀 ). Lastly, in Lightning 2.3 I also added the train mode to the Model Summary (in an extra column).

**Comment by StevenSong** [comment]
Ah, thanks for the reply and the info, I should've done a little more digging into the issue/PR history.

> ```python
> hf_model = AutoModelForCausalLM.from_pretrained("openai-community/gpt2").train()
> ```

This certainly would fix it, I've implemented it my training scripts. 

Thanks again, happy to close this and migrate the conversation to the other thread
